ouroboros-consensus-test/test-consensus/Test/Consensus/Mempool/StateMachine.hs

```diff
@@ -0,0 +1,810 @@
+              }
+  (Action GetSnapshot, GotSnapshot{}) -> model
+  (Action SyncLedger, Void) ->
+    if mtip == tick cfg dbtip
+    then model
+    else
+      let modelTxs'       = fmap fst mtxs
+          (res, _, _, st) = foldTxs cfg zeroTicketNo modelCapacity
+                              (tick cfg dbtip)
+                              modelTxs'
+          filteredTxSeq   =
+            [ (tx, tk) | (tx, tk) <- mtxs
+                          , tx
+                            `elem` [ txForgetValidated vtx
+                                   | MempoolTxAdded vtx <- fst res
+                                   ]
+                     ]
+      in
+        model {
+          modelMempoolTip = (st, True)
+        , modelTxs = filteredTxSeq
+        , modelCapacity = MempoolCapacityBytes 400
+        }
+
+  _ -> error $ "mismatched command "
+             <> show cmd
+             <> " and response "
+             <> show resp
+
+  where
+    Model {
+        modelMempoolTip  = (mtip, reachable)
+      , modelLedgerDBTip = dbtip
+      , modelTxs         = mtxs
+      , modelCapacity
+      , modelSeenStates  = seenStates
+      , modelMempoolWasFilled
+      } = model
+
+{-------------------------------------------------------------------------------
+  Ledger helper functions
+-------------------------------------------------------------------------------}
+
+-- | Apply a list of transactions short-circuiting if the mempool gets full.
+-- Emulates almost exactly the behavior of 'implTryAddTxs'.
+foldTxs ::
+  forall blk.
+  ( LedgerSupportsMempool blk
+  , BasicEnvelopeValidation blk
+  )
+  => LedgerConfig blk
+  -> TicketNo
+  -> MempoolCapacityBytes
+  -> TickedLedgerState blk ValuesMK
+  -> [GenTx blk]
+  -> ( ([MempoolAddTxResult blk], [GenTx blk])
+     , TicketNo
+     , MempoolCapacityBytes
+     , TickedLedgerState blk ValuesMK
+     )
+foldTxs cfg nextTk remainingCap initialState  =
+    go ( []
+       , nextTk
+       , getMempoolCapacityBytes remainingCap
+       , initialState
+       )
+  where
+    go (acc, tk, cap, st) [] = ((reverse acc, [])
+                               , tk
+                               , MempoolCapacityBytes cap
+                               , st
+                               )
+    go (acc, tk, cap, st) txs@(tx:next) =
+      if cap < txInBlockSize tx
+      then ((reverse acc, txs), tk, MempoolCapacityBytes cap, st)
+      else
+        let slot = case getTipSlot st of
+              Origin -> minimumPossibleSlotNo (Proxy @blk)
+              At v   -> v + 1
+        in
+        case runExcept $ applyTx cfg DoNotIntervene slot tx st of
+          Left e           ->
+            go ( MempoolTxRejected tx e:acc
+               , tk
+               , cap
+               , st
+               )
+               next
+          Right (st', vtx) ->
+            go ( MempoolTxAdded vtx:acc
+               , succ tk
+               , cap - txInBlockSize tx
```

Comment:
Here we're calculating the remaining capacity anyway, so maybe we can add this to the model and remove the other two fields.